auth_saml/models/res_users.py

```diff
@@ -0,0 +1,219 @@
+
+            :param provider: saml provider id (int)
+            :param validation: result of validation of access token (dict)
+            :param saml_response: saml parameters response from the IDP
+            :return: user login (str)
+            :raise: openerp.exceptions.AccessDenied if signin failed
+
+            This method can be overridden to add alternative signin methods.
+        """
+        token_osv = self.env['auth_saml.token']
+        saml_uid = validation['user_id']
+
+        user_ids = self.search(
+            [
+                ("saml_uid", "=", saml_uid),
+                ('saml_provider_id', '=', provider),
+            ]
+        )
+
+        if not user_ids:
+            raise openerp.exceptions.AccessDenied()
+
+        # TODO replace assert by proper raise... asserts do not execute in
+        # production code...
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [
+                ('saml_provider_id', '=', provider),
+                ('user_id', '=', user.id),
+            ]
+        )
+
+        if token_ids:
+            token_ids.write(
+                {'saml_access_token': saml_response},
+            )
+        else:
+            token_osv.create(
+                {
+                    'saml_access_token': saml_response,
+                    'saml_provider_id': provider,
+                    'user_id': user.id,
+                },
+            )
+
+        return user.login
+
+    @api.model
+    def auth_saml(self, provider, saml_response):
+
+        validation = self._auth_saml_validate(provider, saml_response)
+
+        # required check
+        if not validation.get('user_id'):
+            raise openerp.exceptions.AccessDenied()
+
+        # retrieve and sign in user
+        login = self._auth_saml_signin(provider, validation, saml_response)
+
+        if not login:
+            raise openerp.exceptions.AccessDenied()
+
+        # return user credentials
+        return self.env.cr.dbname, login, saml_response
+
+    # This method is using the old v7 API because it is called BEFORE the login
+    def check_credentials(self, cr, uid, token):
+        """token can be a password if the user has used the normal form...
+        but we are more interested in the case when they are tokens
+        and the interesting code is inside the except clause
+        """
+        token_osv = self.pool.get('auth_saml.token')
+
+        try:
+            baseuser.check_credentials(
+                self,
+                cr,
+                uid,
+                token
+            )
+
+        except (
+            openerp.exceptions.AccessDenied,
+            passlib.exc.PasswordSizeError,
+        ):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            res = token_osv.search(
```

Comment:
While migrate to odoo9 I had a try using @api.model decorator

and this become something like:
```
token_osv.sudo().search(
    [
        ('user_id', '=', self.env.uid),
        ('saml_access_token', '=', token),
    ]
)
```

it looks working!